flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerTest.java

```diff
@@ -707,6 +706,53 @@ public void testTaskManagerTimeout() throws Exception {
 		}
 	}
 
+	/**
+	 * Tests that idle but not releasable task managers will not be released even if timed out before it can be.
+	 */
+	@Test
+	public void testTaskManagerNotReleasedBeforeItCanBe() throws Exception {
+		final long tmTimeout = 10L;
+
+		final CompletableFuture<InstanceID> releaseFuture = new CompletableFuture<>();
+		final ResourceActions resourceManagerActions = new TestingResourceActionsBuilder()
+			.setReleaseResourceConsumer((instanceID, e) -> releaseFuture.complete(instanceID))
+			.build();
+		final ResourceManagerId resourceManagerId = ResourceManagerId.generate();
+		final ResourceID resourceID = ResourceID.generate();
+
+		final AtomicBoolean canBeReleased = new AtomicBoolean(false);
+		final TaskExecutorGateway taskExecutorGateway = new TestingTaskExecutorGatewayBuilder()
+			.setCanBeReleasedSupplier(canBeReleased::get)
+			.createTestingTaskExecutorGateway();
+		final TaskExecutorConnection taskManagerConnection = new TaskExecutorConnection(resourceID, taskExecutorGateway);
+
+		final SlotID slotId = new SlotID(resourceID, 0);
+		final ResourceProfile resourceProfile = new ResourceProfile(1.0, 1);
+		final SlotStatus slotStatus = new SlotStatus(slotId, resourceProfile);
+		final SlotReport slotReport = new SlotReport(slotStatus);
+
+		final Executor mainThreadExecutor = TestingUtils.defaultExecutor();
+
+		try (SlotManager slotManager = SlotManagerBuilder.newBuilder()
+			.setTaskManagerTimeout(Time.milliseconds(tmTimeout))
+			.build()) {
+
+			slotManager.start(resourceManagerId, mainThreadExecutor, resourceManagerActions);
+
+			mainThreadExecutor.execute(() -> slotManager.registerTaskManager(taskManagerConnection, slotReport));
+
+			// now it can not be released yet
+			canBeReleased.set(false);
+			mainThreadExecutor.execute(slotManager::checkTaskManagerTimeouts);
```

Comment:
shouldn't we wait for this to complete, because as it is now the assertion below might be a no-op (depending on the race condition)